mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java

```diff
@@ -58,38 +63,78 @@ public CassandraAnnotationMapper(CassandraId mailboxId, Session session) {
+            .and(eq(CassandraAnnotationTable.KEY, key.getKey())));
     }
 
     public void insertAnnotation(MailboxAnnotation mailboxAnnotation) {
         Preconditions.checkArgument(!mailboxAnnotation.isNil());
         session.execute(insertInto(CassandraAnnotationTable.TABLE_NAME)
             .value(CassandraAnnotationTable.MAILBOX_ID, mailboxId.asUuid())
-            .value(CassandraAnnotationTable.KEY, mailboxAnnotation.getKey())
+            .value(CassandraAnnotationTable.KEY, mailboxAnnotation.getKey().getKey())
             .value(CassandraAnnotationTable.VALUE, mailboxAnnotation.getValue().get()));
     }
 
     private MailboxAnnotation toAnnotation(Row row) {
-        return MailboxAnnotation.newInstance(row.getString(CassandraAnnotationTable.KEY), row.getString(CassandraAnnotationTable.VALUE));
+        return MailboxAnnotation.newInstance(new MailboxAnnotationKey(row.getString(CassandraAnnotationTable.KEY)),
+            row.getString(CassandraAnnotationTable.VALUE));
     }
 
     private Select.Where getStoredAnnotationsQuery() {
         return select(CassandraAnnotationTable.SELECT_FIELDS)
-                .from(CassandraAnnotationTable.TABLE_NAME)
-                .where(eq(CassandraAnnotationTable.MAILBOX_ID, mailboxId.asUuid()));
+            .from(CassandraAnnotationTable.TABLE_NAME)
+            .where(eq(CassandraAnnotationTable.MAILBOX_ID, mailboxId.asUuid()));
+    }
+
+    private Select.Where getStoredAnnotationsQueryForKeys(Set<MailboxAnnotationKey> keys) {
+        return getStoredAnnotationsQuery().and(in(CassandraAnnotationTable.KEY, keys.stream()
+            .map(MailboxAnnotationKey::getKey)
+            .collect(ImmutableCollectors.toImmutableList())));
+    }
+
+    private Select.Where getStoredAnnotationsQueryLikeKey(String key) {
+        return getStoredAnnotationsQuery()
+            .and(gte(CassandraAnnotationTable.KEY, key))
+            .and(lte(CassandraAnnotationTable.KEY, buildNextKey(key)));
+    }
+    
+    private String buildNextKey(String key) {
+        return key + MailboxAnnotationKey.SLASH_CHARACTER + MAXIMUM_ASCII_CHARACTER;
+    }
+
+    private Stream<MailboxAnnotation> getAnnotationsByKeyWithAllDepth(MailboxAnnotationKey key) {
+        return CassandraUtils.convertToStream(session.execute(getStoredAnnotationsQueryLikeKey(key.getKey())))
+            .map(this::toAnnotation);
+    }
+
+    private Stream<MailboxAnnotation> getAnnotationsByKeyWithOneDepth(MailboxAnnotationKey key) {
+        return CassandraUtils.convertToStream(session.execute(getStoredAnnotationsQueryLikeKey(key.getKey())))
+            .map(this::toAnnotation)
+            .filter(annotation -> isChild(key, annotation));
     }
 
-    private Select.Where getStoredAnnotationsQueryForKeys(Set<String> keys) {
-        return getStoredAnnotationsQuery().and(in(CassandraAnnotationTable.KEY, Lists.newArrayList(keys)));
+    private boolean isChild(MailboxAnnotationKey key, MailboxAnnotation annotation) {
```

Comment:
In this case, it retrieved all the annotations by key (which has prefix = key)
so "isChild" is not really, it just count the number of slash to decide.
But if we put it on MailboxAnnotation, it should check both:
- Annotation's key has prefix as "key"
- Checking the number of slash

So I don't think it belongs to MailboxAnnotation cause we can use the power of storage on this case (Eg: we can direct query from DB)